### PR TITLE
Bufix: Safe duplicate version

### LINKED ISSUE #14962: Duplicated version number created for existing image
### Describe the bug

_Background_
I always create a new version (duplicate) of a base RAW for a different edit so I can trace back any final output that may result. My filenaming system is '<filename>_<version number>' where filename is composed of '<YYYYMMDD_projectname_original camera filename>'.  I use th $VERSION variable in the dt export module to ensure any output image follows this format.  This provides unique identification of every image and its derivatives across my libraries, even when intermediate tiffs are involved in say, focus stacks.

_The issue_
1)  Raw file imported into darktable March 2020 (pre-dt3.4.0)  image-id=3574. Around that time seven duplicates were created for different edits. The versions and the original raw were made into a group (id=3574)

2)  24 July 2023: created new duplicate of the image (id 3574) in dt4.2.1, deleted history from the new version and edited in darkroom. On return to lightable saw that the version number of the new duplicate was '3' - which I knew already existed for that image.  On viewing list of duplicates in Duplicate Manager, there are two entries with version = 3.  I've checked and the new duplicate '3' has overwritten the existing xmp for the previous version 3. 

3)  I accessed backup copies of the dt database and listed info from the images table for the raw image and its versions, as follows:

_pre-dt3.4.0-library.db image table extract from backup archive 2020-12-14_
select id, group_id, film_id,filename,version,max_version from images where filename = "20200325_BonallyWoods_NIK_1413.NEF" order by id;

id    group_id    film_id    filename                        version    max_version
3574    3574    135    20200325_BonallyWoods_NIK_1413.NEF    0    7
3578    3574    135    20200325_BonallyWoods_NIK_1413.NEF    1    7
3582    3574    135    20200325_BonallyWoods_NIK_1413.NEF    2    7
3583    3574    135    20200325_BonallyWoods_NIK_1413.NEF    3    7
3789    3574    135    20200325_BonallyWoods_NIK_1413.NEF    4    7
3790    3574    135    20200325_BonallyWoods_NIK_1413.NEF    5    7
3791    3574    135    20200325_BonallyWoods_NIK_1413.NEF    6    7
3792    3574    135    20200325_BonallyWoods_NIK_1413.NEF    7    7

_library.db-snp-20230722155434 from backup archive BEFORE new duplicate created_
select id, group_id, film_id,filename,version,max_version from images where filename = "20200325_BonallyWoods_NIK_1413.NEF" order by id;

id    group_id    film_id    filename                    version    max_version
3574    3574    135    20200325_BonallyWoods_NIK_1413.NEF    0    7
3578    3574    135    20200325_BonallyWoods_NIK_1413.NEF    1    7
3582    3574    135    20200325_BonallyWoods_NIK_1413.NEF    2    7
3583    3574    135    20200325_BonallyWoods_NIK_1413.NEF    3    7
3789    3574    135    20200325_BonallyWoods_NIK_1413.NEF    4    7
3790    3574    135    20200325_BonallyWoods_NIK_1413.NEF    5    7
3791    3574    135    20200325_BonallyWoods_NIK_1413.NEF    6    7
3792    3574    135    20200325_BonallyWoods_NIK_1413.NEF    7    7

_library.db-snp-20230724163328 from current backup AFTER new duplicate created_
select id, group_id, film_id,filename,version,max_version from images where filename = "20200325_BonallyWoods_NIK_1413.NEF" order by id;

id    group_id    film_id    filename                    version    max_version
3574    3574    135    20200325_BonallyWoods_NIK_1413.NEF    0    0
3578    3574    135    20200325_BonallyWoods_NIK_1413.NEF    1    3
3582    3574    135    20200325_BonallyWoods_NIK_1413.NEF    2    3
3583    3574    135    20200325_BonallyWoods_NIK_1413.NEF    3    3   ****duplicate
3789    3574    135    20200325_BonallyWoods_NIK_1413.NEF    4    3
3790    3574    135    20200325_BonallyWoods_NIK_1413.NEF    5    3
3791    3574    135    20200325_BonallyWoods_NIK_1413.NEF    6    3
3792    3574    135    20200325_BonallyWoods_NIK_1413.NEF    7    3
9744    3574    135    20200325_BonallyWoods_NIK_1413.NEF    3    3    ****duplicate

The dt database has an index, "images_filename_index" ON "images" ( "filename", "version" );  which means that a 'duplicate' is related to a particular image filename. There are two implications from this which may be relevant to the issue:

a) The index is is not unique, it allows duplicates. Therefore the database allows (and cannot trap) insertion of a new image table row with a version number that already exists for the given filename value. The fact I have a duplicate version for a filename suggests the dt code also does not trap this.

b) This index assumes filenames are unique across the whole dt database, which probably is not realistic given how cameras from the same manufacturer can generate common filenames.

While a unique id is given to each imported image by the dt db to ensure images with the same filename are permitted and can be handled, it seems the 'duplicate image' functionality does not recognise this. 

### Steps to reproduce

1) import raw image file into dt
2) create seven duplicates of the raw image using 'Create Duplicate'
3) Edit the duplicates in darkroom
4) At some point in the future, after several dt version updates, create a new duplicate for the raw file in (1)
5) Discard the history from the new duplicate and edit it in darkroom
6) Return to lightable and version number of the new duplicate is 3, which is less than the number of existing duplicates for the image, and there is pre-existing duplicate with version = 3. The xmp file has been overwritten.

### Expected behavior

DT should always allocate a unique version number when creating duplicates of an image.

### Logfile | Screenshot | Screencast

_No response_

### Commit

_No response_

### Where did you install darktable from?

OBS

### darktable version

darktable 4.2.1

### What OS are you using?

Linux

### What is the version of your OS?

Linux Mint 21,Ubuntu 22.04 jammy

### Describe your system?

_No response_

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

_No response_

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

_No response_

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
We can't take for granted that we have an unused version by taking `max_version+1` as
1. other database might also use this image file or
2. duplicates had been removed from and later added to database.

Fixes #14962

### CODE REVIEW

**Comment by jenshannoschwalm** [comment]
@wpferguson any chance for lua problems?

File: src/common/image.c [hunk 1]
```diff
@@ -1148,6 +1148,39 @@ int32_t dt_image_duplicate(const dt_imgid_t imgid)
...
+                                            const int32_t max_version)
+{
+  int32_t safe_max_version = 0;
+  char imgpath[PATH_MAX] = { 0 };
+  char versionpath[PATH_MAX] = { 0 };
+
+  gboolean from_cache = FALSE;
+  dt_image_full_path(imgid, imgpath, sizeof(imgpath), &from_cache);
+
+  for(int32_t version = MAX(1, max_version); version < 100; version++)
```
Comment (ralfbrown) [inline src/common/image.c:1161]: The sidecar files can use up to four digits, so version numbers up to 9999 are possible (though of course extremely unlikely).

Comment (jenshannoschwalm) [reply src/common/image.c:1161]: Oops, I misread the code elsewhere, it's just at least two decimal digits.

**Comment by wpferguson** [comment]
> any chance for lua problems?

Tested against Lua.  No problems found :)

File: src/common/image.c (hunk 1 above)
Comment (jenshannoschwalm) [reply src/common/image.c]: fixed with force push

**Review by TurboGit — APPROVED** [review]
Looks good to me, thanks!

**Comment by TurboGit** [comment]
Need a release note entry for this I think.

**Comment by jenshannoschwalm** [comment]
@TurboGit suggested release note:

Fixing a bug for duplicate xmp sidecars making sure it has not been used before across different databases.

**Comment by TurboGit** [comment]
Thanks!

**Pull request merged**